Finished the test problems.

MeArio opened the pull request:
### Running:

Run the tests in a terminal with:
`dotnet test`

### Dependencies
Project 'ProblemsSolution' has the following package references
   [netstandard2.0]:

| Top-level Package      | Resolved |
| --------------------------- | ------------ |
| NETStandard.Library | 2.0.3        |

Project 'ProblemsSolution.Tests' has the following package references
   [netcoreapp3.0]:

| Top-level Package         | Resolved |
| ------------------------------| ------------ |
| Microsoft.NET.Test.Sdk | 16.2.0      |
| nunit                              | 3.12.0     |
| NUnit3TestAdapter        | 3.13.0     |

MeArio committed:
Initial Setup

MeArio committed:
Added .gitignore and test for Simple Problem #1

MeArio committed:
Added solution for Simple Problem #1

MeArio committed:
updated .gitignore to include .vscode and omnisharp.json

MeArio committed:
Added test for Simple Problem #2 and removed unnecesary code

MeArio committed:
Added solution for Simple Problem #2

MeArio committed:
Added test for DuplicatesInArray method

MeArio committed:
fixed test data for DuplicatesInArrayTest

MeArio committed:
Added solution for Simple Problem #3

alex-damian-negru commented on ProblemsSolution/ProblemsSolution.cs:
```diff
@@ -0,0 +1,60 @@
+        public static int SumArrayEvenNumbers(int[] numbers)
+        {
+            int sum = 0;
+            foreach(int number in numbers)
```
This might be simplified. Ever heard of [map](https://blog.submain.com/csharp-functional-programming/)?

MeArio replied:
I wasn't sure if I was allowed to use LINQ or not so I decided to do it the tedious way. Also not sure how I would do this with map? I'm thinking about filtering even numbers and then reducing with a sum function.

MeArio replied:
> This might be simplified. Ever heard of [map](https://blog.submain.com/csharp-functional-programming/)?

I've simplified it here aefbbf8c1c711595bcbe35054bb1b342993a8d8c

alex-damian-negru commented on ProblemsSolution/ProblemsSolution.cs:
```diff
@@ -0,0 +1,60 @@
+
+            if(indexSum == arraySum)
+            {
+                return false;
```
The ternary operator could cut up a few lines of code here.

MeArio replied:
> The ternary operator could cut up a few lines of code here.

Fixed fb48401177edc8163c9370bff510a8dc81099111.

alex-damian-negru commented on ProblemsSolution.Tests/ProblemsSolutionTest.cs:
```diff
@@ -0,0 +1,33 @@
+        [TestCase (new int[] {1,2,3,4,5}, ExpectedResult=6)]
+        [TestCase (new int[] {12, 22, 1, 3}, ExpectedResult=34)]
+        [TestCase (new int[] {3, 7, 0, 1}, ExpectedResult=0)]
+        public int SumArrayEvenNumbersTest(int[] numbers)
```
Is it possible to add a short description of each test case? One must be able to see quickly what each test is about, so analyzing every input and extract the description from there is not always great.

MeArio replied:
> Is it possible to add a short description of each test case? One must be able to see quickly what each test is about, so analyzing every input and extract the description from there is not always great.

Done 80f307ae3cfcdd0b3f05df8f44780a5638854953.

alex-damian-negru commented on ProblemsSolution/ProblemsSolution.cs:
```diff
@@ -0,0 +1,60 @@
+
+        public static int[] BigestTwoNumbersArray(int[] numbers)
+        {
+            if(numbers.Length == 0) return new int[] {};
```
Why do you need this guard in place?

MeArio replied:
> Why do you need this guard in place?

Because otherwise the method would return an array that contains 2 elements with value of Int32.MinValue.

alex-damian-negru reviewed (commented):
Check this gem out regarding commit messages: https://chris.beams.io/posts/git-commit/

adrianbunea commented on ProblemsSolution.Tests/ProblemsSolutionTest.cs:
```diff
@@ -0,0 +1,33 @@
+        }
+        [TestCase (new int[] {1, 2, 3, 4, 5}, ExpectedResult=false)]
+        [TestCase (new int[] {1, 2, 3, 2, 5, 6}, ExpectedResult=true)]
+        [TestCase (new int[] {2, 3, 3, 7, 8, 5, 4, 7}, ExpectedResult=true)]
```
Add this test case and run the tests again [TestCase (new int[] {1, 1, 3, 5, 5}, ExpectedResult=true)]

MeArio replied:
> Add this test case and run the tests again [TestCase (new int[] {1, 1, 3, 5, 5}, ExpectedResult=true)]

Yeah, my solution doesn't handle that edge case I will have to think of a better one.

alexandru-calinoiu reviewed (commented):
Please update the pr description to provide us with minimum instructions on how to run the code and it's dependencies versions.

MeArio commented:
> Please update the pr description to provide us with minimum instructions on how to run the code and it's dependencies versions.

Done

MeArio committed:
Simplify SumArrayEvenNumbersTest method

MeArio committed:
Add descriptions to each test case and remove 1 redundant test

MeArio committed:
Simplify DuplicatesInArray method conditional

alexandru-calinoiu commented on ProblemsSolution/ProblemsSolution.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        public static int SumArrayEvenNumbers(int[] numbers)
+        {
+            int sum = numbers.Where(number => number % 2 == 0).Aggregate((a, b) => a+b);
```
This can be done nicer,

Can you levreag extension (or whatever mecanism c# has) to create an .even method 
Also you should be able to simply use `.Sum`

MeArio replied:
Done a50351dd8b44e6452baf759fd43fda28b05561e0.

alexandru-calinoiu commented on ProblemsSolution/ProblemsSolution.cs:
```diff
@@ -0,0 +1,50 @@
+            for(int i=1; i <= numbers.Length; i++){
+                indexSum += i;
+                arraySum += numbers[i - 1];
+            }
```
Not the right solution

It will fail for a simple array like: [2, 2, 2]

MeArio replied:
Fixed: 3d9a558d4d79936dcaf2cd33653540b7dbe1d432 not sure if I should have made a Product extension method since it didn't feel needed.

alexandru-calinoiu commented on ProblemsSolution.Tests/ProblemsSolutionTest.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+
+        [TestCase (new int[] {1,2,3,4,5}, ExpectedResult=6, Description="Check for array with even numbers.")]
+        [TestCase (new int[] {3, 7, 0, 1}, ExpectedResult=0, Description="Check for array with even numbers and a zero.")]
```
The description is incorrect.

MeArio replied:
Fixed 3a3cecea212765927f6183cdb4f8273d4c578ee9.

alexandru-calinoiu commented on ProblemsSolution.Tests/ProblemsSolutionTest.cs:
```diff
@@ -0,0 +1,32 @@
+        [TestCase (new int[] {12, 22, 1, 3}, ExpectedResult=new int[] {12, 22}, Description="Check for array with more than 2 elements.")]
+        [TestCase (new int[] {3}, ExpectedResult=new int[] {3}, Description="Check for single element array.")]
+        [TestCase (new int[] {7, 3}, ExpectedResult=new int[] {3, 7}, Description="Check for 2 element array.")]
+        [TestCase (new int[] {}, ExpectedResult=new int[] {}, Description="Check for empty array.")]
```
How about test cases with negative numbers?

MeArio replied:
> How about test cases with negative numbers?

Yeah it seems that when it comes to negative numbers the array order is reversed, will fix.

MeArio replied:
Fixed 45ff35598dc7149787063f58f7e9d84a406b16b2.

alexandru-calinoiu commented on ProblemsSolution.Tests/ProblemsSolutionTest.cs:
```diff
@@ -0,0 +1,32 @@
+        public int[] BigestTwoNumbersArrayTest(int[] numbers)
+        {
+            return ProblemsSolution.ArrayProblemsSolution.BigestTwoNumbersArray(numbers);
+        }
```
Style looks consistent except this missing empty new line, Wonder what happened here?

MeArio replied:
I didn't notice it.

MeArio replied:
Fixed af59a0081a4531bc447a1c0ef6d2af8b1b5da9f9.

alexandru-calinoiu reviewed (commented):
A decent try, please see my comments.

Looks like for the first solution you skiped over the TDD part.

Looking further down over the commits, this is not really TDD, take a look at this guy to see what I was really expecting https://www.youtube.com/watch?v=qkblc5WRn-U

Some more comments on your commits, they don't seem to be consistent
I almost like: Add descriptions to each test case and remove 1 redundant test 
Don't really like (because is not consistent): Added solution for Simple Problem #3
A good readup about commit messages https://chris.beams.io/posts/git-commit/

MeArio committed:
Add new line between test methods for clarity

MeArio committed:
Add Extensions class

MeArio committed:
Add Extensions class

MeArio committed:
Change name of ProblemsSolutionTest.Tests.Test to ArrayProblemsSolutionTest

MeArio committed:
Add test for Even extension method

MeArio committed:
Implement Even method

MeArio committed:
Refactor SumArrayEvenNumbers method

MeArio commented:
> A decent try, please see my comments.
> 
> Looks like for the first solution you skiped over the TDD part.
> 
> Looking further down over the commits, this is not really TDD, take a look at this guy to see what I was really expecting https://www.youtube.com/watch?v=qkblc5WRn-U
> 
> Some more comments on your commits, they don't seem to be consistent
> I almost like: Add descriptions to each test case and remove 1 redundant test
> Don't really like (because is not consistent): Added solution for Simple Problem #3
> A good readup about commit messages https://chris.beams.io/posts/git-commit/

> A decent try, please see my comments.
> 
> Looks like for the first solution you skiped over the TDD part.
> 
> Looking further down over the commits, this is not really TDD, take a look at this guy to see what I was really expecting https://www.youtube.com/watch?v=qkblc5WRn-U
> 
> Some more comments on your commits, they don't seem to be consistent
> I almost like: Add descriptions to each test case and remove 1 redundant test
> Don't really like (because is not consistent): Added solution for Simple Problem #3
> A good readup about commit messages https://chris.beams.io/posts/git-commit/

I will give that video a watch as soon as I have time.

I should have probably separated that commit in 2 different ones.

Yeah I know the solution is useless I've never been good at these type of problems that's why I never got into competitive programming.

Someone else suggested that read on this PR that's why I've been changing my commit message style and why the commit history is inconsistent.

alexandru-calinoiu commented on ProblemsSolution/obj/ProblemsSolution.csproj.nuget.cache:
```diff
@@ -0,0 +1,5 @@
+{
```
I am preatty sure this `obj` folder should have been excluded.

MeArio replied:
Fixed I probably forgot to remove the cached files after fixing the .gitignore.

MeArio committed:
Remove files missed by the initial .gitignore

MeArio committed:
Fix inconsistency in [TestCase] attribute order

MeArio committed:
Add test for ccurrent implementation edge case

MeArio committed:
Fix DuplicatesInArray implementation, solved for "identical sums" edge case

MeArio committed:
Fix [TestCase] description

MeArio committed:
Add regression [TestCase] for negative numbers resulting in a reverse order array

MeArio committed:
Fix BiggestTwoNumbersArray implementation solved for "reverse order array with negative numbers" corner case

alexandru-calinoiu commented:
Thank you, will followup via email.
